### ISSUE TITLE
sensor.home_connect_status (loading)

### ISSUE THREAD

**Issue by zufivan** [ekutner/home-connect-hass#10]
Hi, many thanks for your integration, I like it. 
 I’m currently testing with 5 home connect devices (2 hoven, 1 caffe machine, 1 cooking_hob and 1 dishwasher) but often I have “sensor.home_connect_status” to loading, see the log below:
2022-02-18 16:07:33 DEBUG (MainThread) [home_connect_async.api] HTTP Error 429 - Too Many Requests. Sleeping for 56605 seconds and will retry
Any suggestion?
Many thanks

**Comment by ekutner** [comment]
This happens because of Home Connect's crazy rate limits. With 5 devices there are many calls to the service when loading the integration for the first time. This usually isn't a problem, however, if you have pressed the "Home Connect Refresh" button or restarted the integration before it finished loading that would create even more calls and cause the service to block additional calls for up to 24h.
At this point the only thing you can do is create a new app in the Home Connect website and use its keys, then be careful not to reload the integration or press the refresh button multiple times in a short period of time.

**Comment by zufivan** [comment]
Many thanks for your answer, I will try to create a new app.

**Comment by zufivan** [comment]
Hi ekutner,
This morning I reboot Home Assistant 2 time and now is loading again. 
Maybe you think is possible to create 5 new app on Home Connect website , and configure in your integration to use for each appliance a "client_id" and "client_secret" separately just work around this problem?

**Comment by ekutner** [comment]
It's not possible to do that since each app would then require a separate login, which is not something HA supports.
I'm working on a release that will reduce the number of calls but it will take a few more days. Until then there is not much you can, sorry about that.

**Comment by zufivan** [comment]
many thanks for your support.

**Comment by code-echobase** [comment]
just letting you know that I am also running into the max number of calls during normal use. I have a hob, 2 x ovens, coffee machine, washing machine and dryer. When cooking the integration quickly blows up the max api limit due to the nature of cooking (oven doors, programs, hob turned up and down).

Have tried disabling some sensors to reduce the load as I noticed in the code if the last update was < 60 secs then the integration will update state.

**Comment by ekutner** [comment]
Please try version 0.4.1 it changes the way it fetches data from the cloud service and significantly reduces the number of calls, however, this should have the most impact when HA is restarted. If you still have issues after that then please follow the instructions for adding the log configuration to your configuration.yaml then send me the log after the next time you get blocked.

**Comment by zufivan** [comment]
Hi Ekutner, 
thanks, I installed version 0.4.1 from hacs, after reboot the sensor is loading , I don't know if today with the version 0.4.0 I did a lot of calls. 
Can I send you the full log in private?

**Comment by ekutner** [comment]
0.4.0 and 0.4.1 are the same in that regard, they both use the reduced calls approach. Did you have rate limit issues with 0.4.0 ?

**Comment by zufivan** [comment]
when I installed the 0.4.0 I had loading but I did't change client id and secret, I waited 24 hours and until yesterday was working good. see in attached images, 
![image](https://user-images.githubusercontent.com/47624226/155410943-d3689b47-4273-4956-887f-53289c27365f.png)

**Comment by code-echobase** [comment]
Thanks - loaded 0.4.0 and .1 and not running into this problem any more. Will let you know with the logs if I run into this issue again.

**Comment by zufivan** [comment]
Hi Ekutner, 
just to know, this morning at 7:29 (wake up coffee machine) the status of the sensor.home_connect_status is "ready" now everything works. I will keep you updated if the issue recurs.

**Comment by ekutner** [comment]
@zufivan  @code-echobase 
Sounds good, thanks for the update.

**Comment by zufivan** [comment]
unfortunately yesterday at 18:30 the sensor went into loading again. yesterday I used appliances a lot, could it be the problem? Do you need logs?

**Comment by ekutner** [comment]
@zufivan  Yes, please upload the logs, they will help me figure out what happened.

**Comment by zufivan** [comment]
Can I send to you the full log in private? Witch way?

**Comment by ekutner** [comment]
There shouldn't be anything really sensitive in the log but you can send it
to me on discord if you prefer, using the same username as here on Github.

**Comment by zufivan** [comment]
Hi Ekutner,
sorry I can't add you in discord, no problem, attached the debug log concerning home_connect_alt.

[debug.log](https://github.com/ekutner/home-connect-hass/files/8144257/debug.log)

many thanks

**Comment by ekutner** [comment]
Hi @zufivan, I do see quite a few rate limit errors in the log but even when they happen I'm not seeing enough API calls that would trigger this limit. Are you using the same Home Connect Client ID in more than one place. For example, do you have more than a single Home Assistant setups running, or using the built-in integration together with this one on the same Client ID? If you are doing any of this the please create a dedicated app with a different client ID for every instance of HA and every integration that calls the HC API.

**Comment by zufivan** [comment]
Hi @ekutner, I have only 1 home hassistant setup running.
The “client secret” and “client id” used for home connect built-in integration are associated with another account in home connect app and home connect developer. 
I only keep your integration active and I give you a feed-back in a few days (I'm on vacation).
Thanks.

**Comment by ekutner** [comment]
Thanks, I've reached out to Home Connect support regarding those rate limits. There isn't much more I can do beyond what the recent version is already doing.

**Comment by zufivan** [comment]
Hi Ekutner,
I was on vacation 3 days without using the appliances and the sensor was always on "ready". Today, unfortunately, when using appliances, the problem reoccurred. I did reboot home hassistant after update, and loading appear.  I confirm that I am only using your integration. See log in attachment. 
Many thanks.
[home-assistant.1.log](https://github.com/ekutner/home-connect-hass/files/8179497/home-assistant.1.log)
[home-assistant.log](https://github.com/ekutner/home-connect-hass/files/8179498/home-assistant.log)

**Comment by ekutner** [comment]
These log files seem to be filtered and don't contain all the data that is expected to be in them. How did you get them? You should copy the actual file from /config/home-assistant.log

**Comment by zufivan** [comment]
I filtered in notepad++ for "home_connect" and copied the result, see in attacched full log
[home-assistant.log](https://github.com/ekutner/home-connect-hass/files/8180035/home-assistant.log)
in configuration.yaml I have this debug:
```
logger:
  default: warn
  logs:
    homeassistant.components.lovelace: debug
    home_connect_async: debug
    home_connect_alt: debug
    custom_components.home_connect_alt: debug
```

**Comment by ekutner** [comment]
Thanks for uploading the full log, now all the expected information is there.
Unfortunately, it seems the same as before. Your account is  getting blocked without making too many calls to the service. The limit is supposed to be 50 calls/min but the log between 13:18:20-13:18-30 shows the account was blocked after making only 12 calls. To make it worse, it wasn't blocked for just a few seconds, as it is supposed to be (like happened at 9:05:20) but blocked for 24h. I don't know why this happens. If there was some other component using the same Home Connect app (=same client ID) and the same username that might explain it, but since you say there isn't anything like that then it has to be a problem with the HC service itself. I have already reported this issue to their developers support but they are extremely slow to respond, if they respond at all.
I will let you know as soon as I get a reply from them but until then I'm afraid there is nothing I can do to resolve this issue.

**Comment by code-echobase** [comment]
Here's my logs showing the calls made by the integration. Works for about 12 hours or so before the API limit is reached
[homeconnect.log](https://github.com/ekutner/home-connect-hass/files/8200643/homeconnect.log)
.

**Comment by ekutner** [comment]
@code-echobase 
This is the same problem. The rate limit is supposed to be 50 requests/min but here you were blocked after making only 13 calls. This has already happened several times in the log but in the previous times the block was imposed for just a few seconds while here it was imposed for 24 hours. I'm still waiting to hear back from BSH re. this issue.

**Comment by code-echobase** [comment]
thanks for taking a look. just wondering if the integration is reloading the appliance data when it gets a 404

```
2022-03-07 07:19:18 DEBUG (MainThread) [home_connect_async.appliance] Loaded 7 Options for selected/ConsumerProducts.CoffeeMaker.Program.MyCoffee.MyCoffee2
2022-03-07 07:19:18 DEBUG (MainThread) [home_connect_async.appliance] Loaded data for selected Program
2022-03-07 07:19:18 DEBUG (MainThread) [home_connect_async.appliance] Loaded 26 available Programs
2022-03-07 07:19:18 DEBUG (MainThread) [home_connect_async.callback_registery] Broadcasting event: Events.PROGRAM_SELECTED = None
2022-03-07 07:19:18 DEBUG (MainThread) [home_connect_async.callback_registery] Broadcasting event: Events.DATA_CHANGED = None
2022-03-07 07:19:18 DEBUG (MainThread) [home_connect_async.callback_registery] Broadcasting event: BSH.Common.Root.SelectedProgram = ConsumerProducts.CoffeeMaker.Program.MyCoffee.MyCoffee2
2022-03-07 07:19:18 DEBUG (MainThread) [home_connect_async.callback_registery] Broadcasting event: ConsumerProducts.CoffeeMaker.Option.CoffeeTemperature = ConsumerProducts.CoffeeMaker.EnumType.CoffeeTemperature.94C
2022-03-07 07:19:18 DEBUG (MainThread) [home_connect_async.callback_registery] Broadcasting event: ConsumerProducts.CoffeeMaker.Option.FillQuantity = 150
2022-03-07 07:19:20 DEBUG (MainThread) [home_connect_async.homeconnect] Received event from SSE stream: MessageEvent(type='STATUS', message='STATUS', data='{"items":[{"timestamp":1646637560,"handling":"none","uri":"/api/homeappliances/SIEMENS-CT836LEB6-XXXXXXXXXXX/status/BSH.Common.Status.OperationState","key":"BSH.Common.Status.OperationState","value":"BSH.Common.EnumType.OperationState.ActionRequired","level":"hint"}],"haId":"SIEMENS-CT836LEB6-XXXXXXXXXXX"}', origin='https://api.home-connect.com', last_event_id='SIEMENS-CT836LEB6-XXXXXXXXXXX')
2022-03-07 07:19:20 DEBUG (MainThread) [home_connect_async.appliance] Starting to load appliance data for Coffee Maker (SIEMENS-CT836LEB6-XXXXXXXXXXX)
2022-03-07 07:19:20 DEBUG (MainThread) [home_connect_async.appliance] Loaded 7 Options for selected/ConsumerProducts.CoffeeMaker.Program.MyCoffee.MyCoffee2
2022-03-07 07:19:20 DEBUG (MainThread) [home_connect_async.appliance] Loaded data for selected Program
2022-03-07 07:19:20 DEBUG (MainThread) [home_connect_async.api] API call failed with code=404 error=SDK.Error.NoProgramActive
2022-03-07 07:19:20 DEBUG (MainThread) [home_connect_async.appliance] Failed to load active programs with error code=404 key=SDK.Error.NoProgramActive
2022-03-07 07:19:20 DEBUG (MainThread) [home_connect_async.appliance] Loaded 2 Settings
2022-03-07 07:19:20 DEBUG (MainThread) [home_connect_async.appliance] Loaded 11 Statuses
2022-03-07 07:19:20 DEBUG (MainThread) [home_connect_async.appliance] Loaded 0 Commands
```

and whether this is contributing to the number of calls

**Comment by ekutner** [comment]
yes, every "loaded" message is a single call except for for selected / active programs where there are two messages (one for the options and one for the program) although there is just one call.

**Comment by ekutner** [comment]
@code-echobase  @zufivan 
I got information for the Home Connect support team that there is a second rate limit in place which I wasn't aware of before. This rate limit is for 1000 requests in 24h and is probably the reason for this occasional lock-out. I asked them to increase that limit so customers with many appliances will not have a problem but as on now they are not willing to do so.
I've added a call counter to the log in order to try to better analyze this issue. If it is still happening then please upload a debug log for review.

**Comment by zufivan** [comment]
hi @ekutner, sorry for delay to response,  see in attacched log to analizyze the issue.
[home-assistant.log](https://github.com/ekutner/home-connect-hass/files/8884569/home-assistant.log)
many thanks for your great work.

Regards Ivan

**Comment by ekutner** [comment]
it looks like you are using version 0.5.0-b2 which is an unstable beta version. I suggest you roll back to 0.4.10 for now, until 0.5.0 is released.
